src/ray/gcs/gcs_server/gcs_actor_distribution.cc

```diff
@@ -0,0 +1,227 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/gcs/gcs_server/gcs_actor_distribution.h"
+#include "ray/util/event.h"
+
+namespace ray {
+
+namespace gcs {
+
+GcsActorWorkerAssignment::GcsActorWorkerAssignment(const NodeID &node_id,
+                                                   const ResourceSet &acquired_resources,
+                                                   bool is_shared)
+    : node_id_(node_id), acquired_resources_(acquired_resources), is_shared_(is_shared) {}
+
+const NodeID &GcsActorWorkerAssignment::GetNodeID() const { return node_id_; }
+
+const ResourceSet &GcsActorWorkerAssignment::GetResources() const {
+  return acquired_resources_;
+}
+
+bool GcsActorWorkerAssignment::IsShared() const { return is_shared_; }
+
+GcsBasedActorScheduler::GcsBasedActorScheduler(
+    instrumented_io_context &io_context, gcs::GcsActorTable &gcs_actor_table,
+    const GcsNodeManager &gcs_node_manager, std::shared_ptr<gcs::GcsPubSub> gcs_pub_sub,
+    std::shared_ptr<GcsResourceManager> gcs_resource_manager,
+    std::shared_ptr<GcsResourceScheduler> gcs_resource_scheduler,
+    std::function<void(std::shared_ptr<GcsActor>)> schedule_failure_handler,
+    std::function<void(std::shared_ptr<GcsActor>)> schedule_success_handler,
+    std::shared_ptr<rpc::NodeManagerClientPool> raylet_client_pool,
+    rpc::ClientFactoryFn client_factory)
+    : GcsActorScheduler(io_context, gcs_actor_table, gcs_node_manager, gcs_pub_sub,
+                        schedule_failure_handler, schedule_success_handler,
+                        raylet_client_pool, client_factory),
+      gcs_resource_manager_(std::move(gcs_resource_manager)),
+      gcs_resource_scheduler_(std::move(gcs_resource_scheduler)) {}
+
+NodeID GcsBasedActorScheduler::SelectNode(std::shared_ptr<GcsActor> actor) {
+  // TODO(Chong-Li): Java actors may not need a sole assignment (worker process).
+  bool need_sole_actor_worker_assignment = true;
+  if (auto selected_actor_worker_assignment = SelectOrAllocateActorWorkerAssignment(
+          actor, need_sole_actor_worker_assignment)) {
+    actor->SetActorWorkerAssignment(selected_actor_worker_assignment);
+    return selected_actor_worker_assignment->GetNodeID();
+  }
+  return NodeID::Nil();
+}
+
+std::shared_ptr<GcsActorWorkerAssignment>
```

Comment:
unique_ptr?

Reply:
It is used to work with the shared_ptr in GcsActor (see above comment).

Reply:
Do you mean this one? Should this be a unique_ptr then? (Or should this be an `optional`?)
```
void GcsActor::SetActorWorkerAssignment(
    std::shared_ptr<GcsActorWorkerAssignment> assignment_ptr) {
```

Reply:
The `std::shared_ptr<GcsActorWorkerAssignment>` in `gcs_actor_distribution.cc` have been fixed. But in `class GcsActor`, we can still only use `shared_ptr`, for the same reason above.